### PR TITLE
Train loss

### PR DESCRIPTION
**Pull request by undertherain**
moved training loss inside the model definition

### CODE REVIEW

**Comment by vatai** [comment]
# The idea
Have different head/top on the network based on whether we are doing training or inference. @undertherain created the `Net4Inference` and the `Net4Training` classes. I added the `Net4Both` class. Now when you create a network (of course we are talking **only** about pytorch) you do something like this
```python
import torch
import torch.nn.functional as F
import torchvision.models as models

from ..helpers_torch import Net4Both


def get_kernel(params, unparsed_args):
    net = MyNet()
    return Net4Both(
        params,
        net,
        lambda t: F.softmax(t, dim=-1), # top/head inference
        torch.nn.CrossEntropyLoss(), # top/head for training
    )
```
# Some other "fixes" along the way
## F.softmax depreciation warning
Inference and `F.softmax` complained about depreciation warning, that you must use the `dim=d` parameter... **I think** I made the right call with `dim=-1` - the possible values for `dim=` can be `[-2,..,1]`. I did some debugging and it looked ok (both for 1 and -1, but I guess they are the same in this case). Sadly this has to be done per network... using a `lambda` as shown in the above code example.

## Classification generation
Heh, this was funny... so apparently the code for generating the `Y`s for classification looked like this:
```python
    Y = np.random.random(shape[:2]).astype(np.int64)
```
Now if you think about it... this creates a bunch of numbers using `np.random.random()`... which I think generates numbers in the `(0,1)` interval... so after `.astype(np.int64)`... you get a bunch of `0`s :smile_cat: 
So I replaced it with this:
```python
    Y = np.random.randint(0, num_cls, shape[:2])
```
this change can be found in `benchmarker/util/data/synthetic/helpers.py`.

**Comment by vatai** [comment]
## One more fix I forgot
The `img_X_cls.py` synthetic data files didn't specify the number of classes it generated, so they were renamed to `img_x_cls_Y.py` (`X` is the size of the image, `Y` is the number of classes generated). So the `data.py` files needed to be updated.
